File: backend/app/routers/entries.py
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db.database import get_db
from app.models.user import User
from app.models.entry import Entry
from app.schemas.entry import EntryStart, EntryUpdate, EntryOut
from app.services.entry_service import (
    get_active_entry,
    start_entry,
    stop_entry,
    get_entries_by_date,
    get_entries_by_range,
    update_entry,
    delete_entry,
)
from app.routers.auth import get_current_user
from app.config import TIMEZONE
import zoneinfo
from datetime import datetime

router = APIRouter(prefix="/entries", tags=["entries"])


def _entry_to_out(entry) -> EntryOut:
    duration = None
    if entry.ended_at:
        duration = int((entry.ended_at - entry.started_at).total_seconds())
    return EntryOut(
        id=entry.id,
        user_id=entry.user_id,
        label=entry.label,
        started_at=entry.started_at,
        ended_at=entry.ended_at,
        duration_seconds=duration,
        created_at=entry.created_at,
    )


@router.post("/start", response_model=EntryOut)
def start(payload: EntryStart, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    entry = start_entry(db, current_user.id, payload.label)
    return _entry_to_out(entry)


@router.post("/stop", response_model=EntryOut)
def stop(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    entry = stop_entry(db, current_user.id)
    if entry is None:
        raise HTTPException(status_code=404, detail="No active entry")
    return _entry_to_out(entry)


@router.get("/active", response_model=EntryOut | None)
def active(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    entry = get_active_entry(db, current_user.id)
    if entry is None:
        return None
    return _entry_to_out(entry)


@router.get("", response_model=list[EntryOut])
def list_entries(
    date: str | None = None,
    from_: str | None = None,
    to: str | None = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if date:
        entries = get_entries_by_date(db, current_user.id, date)
    elif from_ and to:
        from app.config import TIMEZONE
        import zoneinfo
        from datetime import datetime as dt_mod
        tz = zoneinfo.ZoneInfo(TIMEZONE)
        from_dt = dt_mod.strptime(from_, "%Y-%m-%d").replace(tzinfo=tz)
        to_dt = dt_mod.strptime(to, "%Y-%m-%d").replace(hour=23, minute=59, second=59, tzinfo=tz)
        entries = get_entries_by_range(db, current_user.id, from_dt, to_dt)
    else:
        entries = []
    return [_entry_to_out(e) for e in entries]


@router.patch("/{entry_id}", response_model=EntryOut)
def patch_entry(entry_id: int, payload: EntryUpdate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    data = payload.model_dump(exclude_unset=True)
    entry = update_entry(db, entry_id, current_user.id, data)
    if entry is None:
        raise HTTPException(status_code=404, detail="Entry not found")
    return _entry_to_out(entry)


@router.delete("/{entry_id}", status_code=204)
def remove_entry(entry_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    deleted = delete_entry(db, entry_id, current_user.id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Entry not found")

File: backend/src/routes/entries.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { entryStartSchema, entryUpdateSchema } from "../schemas/entry.js";
import type { EntryOut } from "../schemas/entry.js";
import {
  getActiveEntry,
  startEntry,
  stopEntry,
  getEntriesByDate,
  getEntriesByRange,
  updateEntry,
  deleteEntry,
} from "../services/entry-service.js";

export const entriesRouter = new Hono();

function toEntryOut(row: { id: number; user_id: number; label: string; started_at: string; ended_at: string | null; created_at: string }): EntryOut {
  let duration: number | null = null;
  if (row.ended_at) {
    duration = Math.floor((new Date(row.ended_at).getTime() - new Date(row.started_at).getTime()) / 1000);
  }
  return {
    id: row.id,
    user_id: row.user_id,
    label: row.label,
    started_at: row.started_at,
    ended_at: row.ended_at,
    duration_seconds: duration,
    created_at: row.created_at,
  };
}

entriesRouter.post("/start", zValidator("json", entryStartSchema), async (c) => {
  const user = c.get("user");
  const { label } = c.req.valid("json");
  const entry = startEntry(user.id, label);
  return c.json(toEntryOut(entry));
});

entriesRouter.post("/stop", async (c) => {
  const user = c.get("user");
  const entry = stopEntry(user.id);
  if (!entry) {
    return c.json({ detail: "No active entry" }, 404);
  }
  return c.json(toEntryOut(entry));
});

entriesRouter.get("/active", async (c) => {
  const user = c.get("user");
  const entry = getActiveEntry(user.id);
  if (!entry) return c.json(null);
  return c.json(toEntryOut(entry));
});

entriesRouter.get("/", async (c) => {
  const user = c.get("user");
  const date = c.req.query("date");
  const from_ = c.req.query("from");
  const to = c.req.query("to");

  let entries;
  if (date) {
    entries = getEntriesByDate(user.id, date);
  } else if (from_ && to) {
    const fromDt = new Date(`${from_}T00:00:00`).toISOString();
    const toDt = new Date(`${to}T23:59:59`).toISOString();
    entries = getEntriesByRange(user.id, fromDt, toDt);
  } else {
    entries = [];
  }
  return c.json(entries.map(toEntryOut));
});

entriesRouter.patch("/:entryId", zValidator("json", entryUpdateSchema), async (c) => {
  const user = c.get("user");
  const entryId = parseInt(c.req.param("entryId"), 10);
  const data = c.req.valid("json");
  const entry = updateEntry(entryId, user.id, data as Record<string, unknown>);
  if (!entry) {
    return c.json({ detail: "Entry not found" }, 404);
  }
  return c.json(toEntryOut(entry));
});

entriesRouter.delete("/:entryId", async (c) => {
  const user = c.get("user");
  const entryId = parseInt(c.req.param("entryId"), 10);
  const deleted = deleteEntry(entryId, user.id);
  if (!deleted) {
    return c.json({ detail: "Entry not found" }, 404);
  }
  return c.body(null, 204);
});